### PR TITLE
Upgrade babel by switching to swc

### PR DESCRIPTION
**Pull request by faultyserver**
# Contributing to the Donation Tracker

First of all, thank you for taking the time to contribute!

Please fill out the template below and check the following boxes:

- [ ] I've added tests or modified existing tests for the change.
- [x] I've humanly end-to-end tested the change by running an instance of the tracker.

### Description of the Change

Our version of babel (7.5.5) was at least on the same major version, but still relatively far behind latest (7.21.4). Partly because of that, our plugin set was also missing a lot of features that we _could_ use in modern typescript, like the null coalescing operator (`maybeUndefined ?? defaultValue;`). In debating how to upgrade babel there are a few points to think about:

- Babel doesn't handle typescript natively and uses a plugin. This is fine, but it's a separate dependency that could mismatch somewhere.
- Babel syntax features are complicated. There's `preset-react` that handles JSX and some other things, `preset-env` that...does a lot but it's not always clear what's there, which then leads to `plugin-proposal-*` and `plugin-syntax-*` packages getting added to opt in to individual features. All more configuration that eventually gets out of date and doesn't match the typescript world that we're already writing.
- There are actual competitors now for transpiling source code. It used to just be `ts-loader`, which would use `tsc` itself as the transpiler and was fairly slow because in inherently did more work than `babel`. But now there are alternatives written in native languages like `swc` and `esbuild`, which prove to be [more than an order of magnitude faster](https://swc.rs/) while providing most of the same feature set.

With that, we now have a choice to upgrade. We can stick with Babel or switch to one of the new toolchains. In this PR, I chose to use `swc`, primarily because it is gaining more support from a larger ecosystem (It's Vercel-backed and adopted by many large companies already) and has a more open ecosystem and friendly configuration. The benefits of swc include:

- Faster builds, both in development and production. This is relatively minimal for us right now because we don't actually have that much javascript to bundle. As such, most of the time in webpack is spent on intermediate processes, so our actual benefit is only about 5% (a production build goes from 6500ms to 6000ms on my machine, dev-server sees more gains from 3200ms to 2400ms and has slightly faster recompiles). As we write more javascript, this difference will only get larger as webpack itself becomes negligible.
- Native typescript integration. We'll have to keep swc up-to-date to get new some typescript _syntax_ features, but that's all and is fairly rare in typescript nowadays.
- Less configuration. There's a configuration object in the webpack config, but it simply enables typescript and react-refresh. Everything else is built in natively, so no more plugin configuration or preset management.
- Obscurities of core-js and others are removed. We don't need `@babel/runtime` or to keep up a `core-js` dependency on our own, that's all handled within swc.
- (Eventually) a much faster minifier. Using `TerserPlugin.swcMinify` would reduce our production build times by more than 50% (from 6500ms to 2100ms), but currently the bundle sizes are much larger, so I'm not enabling that for now.
- (Eventually) faster tests. `@swc/jest` will let us run jest tests using the native compiler rather than pre-bundling or waiting on babel each time. This depends on migrating to jest first, though.

All said, I think swc is a promising direction to go in, improves dev times, and simplifies our configuration. Since we would already be updating babel anyway and wanting to get all of the plugins up to date, this also feels like the right time to make a move.

**NOTE:** This PR also includes changing from the obsolete `react-hot-loader` to the newer React Fast Refresh, which lets pages hot-reload in place and is officially supported by React (hot-loader was effectively deprecated more than a year ago). It's included here only because the webpack loader config was being changed already, and swc's hot-reloading was easy to enable.

### Verification Process

Confirmed that both development and production builds still work, didn't bloat the bundle, and ran faster.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -5,7 +5,6 @@
     "es6": true
   },
   "extends": ["eslint:recommended", "plugin:react/recommended", "plugin:@typescript-eslint/recommended"],
-  "parser": "@babel/eslint-parser",
```
Comment (faultyserver) [inline .eslintrc:8]: I don't think we actually needed this in the first place for our JS. Lint still runs perfectly well without a custom parser here, even managing to understand JSX and other problems. See the unrelated lint fix below for an example.

File: bundles/admin/scheduleEditor/dragDrop/emptyTableDropTarget.js [hunk 2]
```diff
@@ -4,14 +4,14 @@ import { DropTarget } from 'react-dnd';
 
 class EmptyTableDropTarget extends React.Component {
   render() {
-    const { isOver, canDrop, connectDropTarget } = this.props;
+    const { isOver, canDrop, connectDropTarget, elementType: Element } = this.props;
     return connectDropTarget(
-      <this.props.elementType
+      <Element
```
Comment (faultyserver) [inline bundles/admin/scheduleEditor/dragDrop/emptyTableDropTarget.js:9]: This is the unrelated lint fix. ESLint complained that `this` does not exist at the point of the JSX opener.

Most likely this was fine before because babel transformed the `this.props` into a different scope where it _was_ available, but without the transform it has to be understood in place.

This is the only thing that came up when it ran, so I'm pretty confident that we will be okay, especially because our code will be TypeScript going forward anyway.

File: package.json [hunk 3]
```diff
@@ -144,10 +131,11 @@
     "prettier": "^2.1.2",
     "process": "^0.11.10",
     "puppeteer": "^2.0.0",
-    "react-hot-loader": "^4.12.15",
+    "react-refresh": "^0.14.0",
```
Comment (faultyserver) [inline package.json:134]: Even though we don't use this directly, we do actually have to directly depend on it. Similar to the node polyfills from the webpack 5 upgrade, it's just inherent in the tooling and part of working with stricter node_modules structures.

File: webpack.config.js [hunk 4]
```diff
@@ -34,7 +35,29 @@ module.exports = {
...
+                },
+                parser: {
+                  syntax: 'typescript',
+                  tsx: true,
+                },
+                loose: false,
+                transform: {
+                  react: {
+                    refresh: !PROD,
+                    runtime: 'classic',
```
Comment (faultyserver) [inline webpack.config.js:54]: Once we upgrade past React 16, we'll be able to use `runtime: 'automatic'`, which should reduce bundle sizes and use `react/jsx-runtime` as the modern approach.

File: webpack.config.js [hunk 5]
```diff
@@ -108,11 +130,6 @@ module.exports = {
     minimizer: [
       new TerserPlugin({
         parallel: true,
-        terserOptions: {
-          output: {
-            comments: /@license/i,
-          },
-        },
```
Comment (faultyserver) [inline webpack.config.js:115]: `output` was already a deprecated option, but both new versions of terser and swc itself will default to including license comments in the output, so this removal is a no-op change.

File: webpack.config.js [hunk 6]
```diff
@@ -142,6 +160,8 @@ module.exports = {
     new webpack.EnvironmentPlugin({
       NODE_ENV: 'development',
     }),
+    !PROD && new ReactRefreshWebpackPlugin(),
+    new webpack.ProgressPlugin(),
```
Comment (faultyserver) [inline webpack.config.js:164]: Adding the ProgressPlugin primarily to check performance.

**Review by uraniumanchor — DISMISSED** [review]
